src/main/java/sorald/sonar/SonarLintEngine.java

```diff
@@ -0,0 +1,232 @@
+
+    // We need to reinitialise it before starting analysis of any source files on any rules.
+    private AnalysisEngine analysisEngine;
+
+    private SonarLintEngine() {
+        super(null);
+        setLogging(null);
+
+        this.analysisEngine =
+                new AnalysisEngine(analysisGlobalConfig, pluginInstancesRepository, null);
+    }
+
+    private static Path getOrDownloadSonarJavaPlugin() {
+        File cacheDirectory = FileUtils.getCacheDir();
+        String sonarJavaPluginFileName =
+                cacheDirectory + System.getProperty("file.separator") + SONAR_JAVA_PLUGIN_JAR_NAME;
+        File sonarJavaPlugin = new File(sonarJavaPluginFileName);
+        if (sonarJavaPlugin.exists()) {
+            return sonarJavaPlugin.toPath();
+        }
+
+        try {
+            InputStream inputStream = new URL(SONAR_JAVA_PLUGIN_URL).openStream();
+            Files.copy(
+                    inputStream,
+                    Paths.get(sonarJavaPluginFileName),
+                    StandardCopyOption.REPLACE_EXISTING);
+            return new File(sonarJavaPluginFileName).toPath();
+        } catch (IOException ignore) {
+            throw new RuntimeException("Could not download Sonar Java plugin"); // NOSONAR:S112
+        }
+    }
+
+    private static StandaloneGlobalConfiguration buildGlobalConfig() {
+        return StandaloneGlobalConfiguration.builder()
+                .addPlugin(sonarJavaPlugin)
+                .addEnabledLanguage(Language.JAVA)
+                .build();
+    }
+
+    private static PluginInstancesRepositoryWhichCannotBeClosed createPluginInstancesRepository() {
+        var config =
+                new Configuration(
+                        globalConfig.getPluginPaths(),
+                        globalConfig.getEnabledLanguages(),
+                        Optional.ofNullable(globalConfig.getNodeJsVersion()));
+        return new PluginInstancesRepositoryWhichCannotBeClosed(config);
+    }
+
+    private static Map<String, SonarLintRuleDefinition> computeAllRulesDefinitionsByKey() {
+        return loadPluginMetadata(
+                pluginInstancesRepository, globalConfig.getEnabledLanguages(), false);
+    }
+
+    private static AnalysisEngineConfiguration buildAnalysisEngineConfiguration() {
+        return AnalysisEngineConfiguration.builder()
+                .addEnabledLanguages(globalConfig.getEnabledLanguages())
+                .setClientPid(globalConfig.getClientPid())
+                .setExtraProperties(globalConfig.extraProperties())
+                .setWorkDir(globalConfig.getWorkDir())
+                .setModulesProvider(globalConfig.getModulesProvider())
+                .build();
+    }
+
+    /** Get or creates the one and only instance of this class. */
+    public static SonarLintEngine getInstance() {
+        if (theOnlyInstance == null) {
+            theOnlyInstance = new SonarLintEngine();
+        } else {
+            theOnlyInstance.recreateAnalysisEngine();
```

Comment:
Per our discussion, we do not need to `recreateAnalysisEngine()` every time we call `getInstance()`.